conwayste/src/ui/button.rs

```diff
@@ -0,0 +1,187 @@
+    ///
+    /// ```rust
+    /// use ggez::graphics::Font;
+    /// use ui::Button;
+    ///
+    /// fn new(ctx: &mut Context) -> GameResult<MainState> {
+    ///     let font = Rc::new(Font::default());
+    ///     let b = Button::new(ctx, WidgetID::TestButton1,
+    ///             UIAction::PrintHelloWorld,
+    ///             WidgetID::TestButton1Label,
+    ///             font,
+    ///             "TestButton")
+    ///         .label_color(Color::from(css::DARKCYAN))
+    ///         .button_color(Color::from(css::WHITE));
+    ///
+    ///     b.draw(ctx)?;
+    /// }
+    /// ```
+    ///
+    pub fn new(ctx: &mut Context, widget_id: WidgetID, action: UIAction, label_id: WidgetID,
+        font: Rc<Font>, button_text: String) -> Self
+    {
+        let label_position = Point2::new(0.0, 0.0); // label positioning defined an offset to button origin after centering
+        let label = Label::new(ctx, label_id, font, button_text, color_with_alpha(css::WHITE, 0.1), label_position);
+        let label_dims = label.size();
+
+        let dimensions = Rect::new(30.0, 20.0, label_dims.w + BUTTON_LABEL_PADDING_W, label_dims.h + BUTTON_LABEL_PADDING_H);
+
+        let mut b = Button {
+            id: widget_id,
+            label: label,
+            button_color: color_with_alpha(css::DARKCYAN, 0.8),
+            draw_mode: DrawMode::fill(),
+            dimensions: dimensions,
+            hover: false,
+            borderless: false,
+            action: action,
+        };
+        b.center_label_text();
+        b
+    }
+
+    /// Sets the color of the Button's text to the specified ggez `Color`
+    pub fn label_color(mut self, color: Color) -> Self {
+        self.label.set_color(color);
+        self
+    }
+
+    /// Sets the color of the button to the specified ggez `Color`
+    pub fn button_color(mut self, color: Color) -> Self {
+        self.button_color = color;
+        self
+    }
+
+    fn center_label_text(&mut self) {
+        let text_dims = self.label.size();
+        let tmp_label_rect = Rect::new(self.dimensions.x, self.dimensions.y, text_dims.w, text_dims.h);
+        let label_center_point = center(&tmp_label_rect);
+        let button_center = center(&self.dimensions);
+
+        self.label.set_size(Rect::new(self.dimensions.x + (button_center.x - label_center_point.x),
+            self.dimensions.y + (button_center.y - label_center_point.y),
+            text_dims.w,
+            text_dims.h));
+    }
+}
+
+impl Widget for Button {
+    fn id(&self) -> WidgetID {
+        self.id
+    }
+
+    fn on_hover(&mut self, point: &Point2<f32>) {
+        self.hover = within_widget(point, &self.dimensions);
+        // if self.hover {
+        //     println!("Hovering over Button, \"{}\" {:?}", self.label.text, point);
+        // }
+    }
+
+    fn on_click(&mut self, _point: &Point2<f32>) -> Option<(WidgetID, UIAction)>
+    {
+        let hover = self.hover;
+        self.hover = false;
+
+        if hover {
```

Comment:
Why does it matter if we were hovering?

Reply:
The hover state check occurs first, prior to any click checking (see `client.rs`, `update_main_menu()` for an example). Hovering ensures the point for the current frame is bounded for a widget.

If the button is part of a `Pane`, then the `Pane` will forward the `on_click()` to its children. Each child should ensure it's in the proper state before confirming the click; the first child producing `Some()` is what the click will register against.